# Evaluate and Clean Sales Data from a E-commerce Company in the UK

## 1. Analysis objectives

The objective of this analysis project is to dig out the popular product according to the sales data. Therefore, the company can create a decent marketing strategy to improve sales revenue.

## 2. Basic Info

This dataset notes down every order detail of a online retail company in the UK from December 1st 2010 to December 9th 2011, covering the company's business data in different countries. This company mainly sells gifts that satisfying numerous context, including birthday gifts, anniversary gifts, Chirstmas gifts, etc. The main clientele of this company includes wholesalers, which takes up bigger proportion, and individual buyers.

The meanings of each column:
- `InvoiceNo`: Invoice number, a 6-digit number as the unique label of each order. If this code starts with 'C', it means this order is canceled.
- `StockCode`: Product code, a 5-digit number, as a unique label of each product.
- `Description`: The name of product.
- `InvoiceDate`: The issue date of invoice, stands for the ordering date and time.
- `UnitPrice`: Unit price, currency is pound(£).
- `CustomerID`: Customer ID, a 5-digit number, as a unique label of each customer.
- `Country`: Country name, the residence country of each customer.

## 3. Read the data

In [1]:
import pandas as pd

original_data = pd.read_csv('e_commerce.csv')
original_data.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,12/1/2010 8:26,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,12/1/2010 8:26,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,12/1/2010 8:26,3.39,17850.0,United Kingdom


## 4. Evaluation

In [2]:
original_data.sample(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
131527,547568,22949,36 DOILIES DOLLY GIRL,12,3/24/2011 9:50,1.45,13341.0,United Kingdom
484529,577532,17012D,ORIGAMI ROSE INCENSE/CANDLE SET,3,11/20/2011 14:01,0.85,16666.0,United Kingdom
114045,546007,22170,PICTURE FRAME WOOD TRIPLE PORTRAIT,1,3/8/2011 15:34,13.29,NaN,United Kingdom
17921,537781,22294,HEART FILIGREE DOVE SMALL,2,12/8/2010 12:46,1.25,17341.0,United Kingdom
703,536423,21481,FAWN BLUE HOT WATER BOTTLE,6,12/1/2010 12:08,2.95,18085.0,United Kingdom
392501,570693,22417,PACK OF 60 SPACEBOY CAKE CASES,4,10/11/2011 16:50,0.55,16161.0,United Kingdom
442810,574682,22629,SPACEBOY LUNCH BOX,1,11/6/2011 12:44,1.95,14963.0,United Kingdom
413377,572309,35651,VINTAGE BEAD PINK SCARF,4,10/23/2011 15:29,1.65,14432.0,United Kingdom
255942,559418,20724,RED RETROSPOT CHARLOTTE BAG,10,7/8/2011 11:36,0.85,12681.0,France
321059,565122,23293,SET OF 12 FAIRY CAKE BAKING CASES,1,9/1/2011 11:48,0.83,13069.0,United Kingdom


### 4.1 Evaluate if the data is clean

In [3]:
original_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


##### According to the output, there are 541909 observation values. There are null values under `Description` and `CustomerID`.
##### Moreover, the data type of `InvoiceDate` should be date and time, while `CustomerID` should be strings. Data type transformation should be processed later.

### 4.2 Evaluate the missing values

In [4]:
original_data[original_data['Description'].isnull()]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,12/1/2010 11:52,0.0,NaN,United Kingdom
1970,536545,21134,NaN,1,12/1/2010 14:32,0.0,NaN,United Kingdom
1971,536546,22145,NaN,1,12/1/2010 14:33,0.0,NaN,United Kingdom
1972,536547,37509,NaN,1,12/1/2010 14:33,0.0,NaN,United Kingdom
1987,536549,85226A,NaN,1,12/1/2010 14:34,0.0,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
535322,581199,84581,NaN,-2,12/7/2011 18:26,0.0,NaN,United Kingdom
535326,581203,23406,NaN,15,12/7/2011 18:31,0.0,NaN,United Kingdom
535332,581209,21620,NaN,6,12/7/2011 18:35,0.0,NaN,United Kingdom
536981,581234,72817,NaN,27,12/8/2011 10:33,0.0,NaN,United Kingdom


##### There are 1454 missing values in `Description`.
##### From the output, the values that miss `Description` value show a 0 in `UnitPrice`. For testing purpose, an extra filtering condition is added in order to confirm if all rows missing `Description` values has a '0' in `UnitPrice`.

In [5]:
original_data[(original_data['Description'].isnull()) & (original_data['UnitPrice'] != 0)]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country


##### The output shows 0, meaning that the rows missing `Description` values do not contain valid `UnitPrice` values.
##### `Description` and `UnitPrice` are two variables that will contribute heavily to the following analysis process. If they are both missing/invalid, the data piece no longer perform its validity, hence it should be deleted later.

##### `CustomerID`contains null values too, hence we extract the null value first.

In [6]:
original_data[original_data['CustomerID'].isnull()]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,12/1/2010 11:52,0.00,NaN,United Kingdom
1443,536544,21773,DECORATIVE ROSE BATHROOM BOTTLE,1,12/1/2010 14:32,2.51,NaN,United Kingdom
1444,536544,21774,DECORATIVE CATS BATHROOM BOTTLE,2,12/1/2010 14:32,2.51,NaN,United Kingdom
1445,536544,21786,POLKADOT RAIN HAT,4,12/1/2010 14:32,0.85,NaN,United Kingdom
1446,536544,21787,RAIN PONCHO RETROSPOT,2,12/1/2010 14:32,1.66,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
541536,581498,85099B,JUMBO BAG RED RETROSPOT,5,12/9/2011 10:26,4.13,NaN,United Kingdom
541537,581498,85099C,JUMBO BAG BAROQUE BLACK WHITE,4,12/9/2011 10:26,4.13,NaN,United Kingdom
541538,581498,85150,LADIES & GENTLEMEN METAL SIGN,1,12/9/2011 10:26,4.96,NaN,United Kingdom
541539,581498,85174,S/4 CACTI CANDLES,1,12/9/2011 10:26,10.79,NaN,United Kingdom


##### `CustomerID` means customer labels, which is not a crucial variable in the later analysis process. Besides, according to the output, some data pieces that miss `CustomerID` value are still valid, hence we keep those observation values. 

### 4.3 Evaluate the duplicate values

##### According to the meaning of each variable, `InvoiceNo`, `StockCode` and `CustomerID` are unique labels. However, a single order may contain numerous product, hence `InvoiceNo` is allowed to have duplicate values; different orders may have same product, hence `StockCode` is allowed to have duplicate values; a single customer may place more than one order and purchase more than one product, hence `CustomerID` is allowed to have duplicate value.
##### Then, for this dataset, we do not need to evaluate the duplicate value.

### 4.4 Evaluate inconsistant data

##### Inconsistancy is possible in `Country` variable, we need to check if different values means the same country.

In [7]:
original_data['Country'].value_counts()

Country
United Kingdom          495266
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
China                      288
Singapore                  229
USA                        218
UK                         211
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United States               73
United Arab Emirates        68


##### From the variable `Country`, 'USA' and 'United States' both stands for America, while 'United Kingdom', 'UK' and 'U.K.' all stands for United Kingdom. Hence, we need to keep only one value for each country.

### 4.5 Evaluate invalid and wrong data

In [8]:
original_data.describe()

,Quantity,UnitPrice,CustomerID
count,541909.000000,541909.000000,406829.000000
mean,9.552250,4.611114,15287.690570
std,218.081158,96.759853,1713.600303
min,-80995.000000,-11062.060000,12346.000000
25%,1.000000,1.250000,13953.000000
50%,3.000000,2.080000,15152.000000
75%,10.000000,4.130000,16791.000000
max,80995.000000,38970.000000,18287.000000


##### From the results, the variable 'Quantity' and 'UnitPrice' contain negative values, which will impact the upcoming analysis process. Hence, we will extract the negative values in 'Quantity' and interpret their meanings.

In [9]:
original_data[original_data['Quantity'] < 0 ]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
141,C536379,D,Discount,-1,12/1/2010 9:41,27.50,14527.0,United Kingdom
154,C536383,35004C,SET OF 3 COLOURED FLYING DUCKS,-1,12/1/2010 9:49,4.65,15311.0,United Kingdom
235,C536391,22556,PLASTERS IN TIN CIRCUS PARADE,-12,12/1/2010 10:24,1.65,17548.0,United Kingdom
236,C536391,21984,PACK OF 12 PINK PAISLEY TISSUES,-24,12/1/2010 10:24,0.29,17548.0,United Kingdom
237,C536391,21983,PACK OF 12 BLUE PAISLEY TISSUES,-24,12/1/2010 10:24,0.29,17548.0,United Kingdom
...,...,...,...,...,...,...,...,...
540449,C581490,23144,ZINC T-LIGHT HOLDER STARS SMALL,-11,12/9/2011 9:57,0.83,14397.0,United Kingdom
541541,C581499,M,Manual,-1,12/9/2011 10:28,224.69,15498.0,United Kingdom
541715,C581568,21258,VICTORIAN SEWING BOX LARGE,-5,12/9/2011 11:57,10.95,15311.0,United Kingdom
541716,C581569,84978,HANGING HEART JAR T-LIGHT HOLDER,-1,12/9/2011 11:58,1.25,17315.0,United Kingdom


##### From the extract result, we can see the negative value in 'Quantity' comes with a 'InvoiceNo' that starts with 'C', which means a canceled order. 
##### But this inference is not necessarily correct. For the purpose of testing, we will add a filter condition to see if there is some negative quantities that do not show an invoice number starts with 'C.'

In [10]:
original_data[(original_data['Quantity'] < 0) & (original_data['InvoiceNo'].str[0] != 'C') ]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
2406,536589,21777,NaN,-10,12/1/2010 16:50,0.0,NaN,United Kingdom
4347,536764,84952C,NaN,-38,12/2/2010 14:42,0.0,NaN,United Kingdom
7188,536996,22712,NaN,-20,12/3/2010 15:30,0.0,NaN,United Kingdom
7189,536997,22028,NaN,-20,12/3/2010 15:30,0.0,NaN,United Kingdom
7190,536998,85067,NaN,-6,12/3/2010 15:30,0.0,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
535333,581210,23395,check,-26,12/7/2011 18:36,0.0,NaN,United Kingdom
535335,581212,22578,lost,-1050,12/7/2011 18:38,0.0,NaN,United Kingdom
535336,581213,22576,check,-30,12/7/2011 18:38,0.0,NaN,United Kingdom
536908,581226,23090,missing,-338,12/8/2011 9:56,0.0,NaN,United Kingdom


##### The inspection above is proved correct. 
##### However, from the results above, the filtered values come with `UnitPrice` as 0. Therefore, a condition related to `UnitPrice` is added to test.

In [11]:
original_data[(original_data['Quantity'] < 0) & (original_data['InvoiceNo'].str[0] != 'C') & (original_data['UnitPrice'] != 0) ]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country


##### According to the output, the inspection got proved correct. When `Quantity` is negarive, the value satisfy one of the following conditions:
- `InvoiceNo` starts with 'C', which stands for a canceled order.
- `UnitPrice` is 0, meaning that the price is 0 pound.
##### Those data points are not valid, hence no sales contribution. They should not be included in the following analysis process. So, we will delete the rows with negative `Quantity` values.

##### Next, we will extract the rows with a negative `UnitPrice`.

In [12]:
original_data[original_data['UnitPrice'] < 0]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
299983,A563186,B,Adjust bad debt,1,8/12/2011 14:51,-11062.06,NaN,United Kingdom
299984,A563187,B,Adjust bad debt,1,8/12/2011 14:52,-11062.06,NaN,United Kingdom


##### From the output above, the rows with negative`UnitPrice` shows a description of /Adjust bad debt/, not a actual trading order. Therefore, these could be deleted before the analysis process.

## 5. Data Cleaning

##### From the results of the previous evaluation part, the data cleaning includes the following steps:
- Transfer the data type of `InvoiceDate` to date and time.
- Transfer the data type of `CustomerID` to strings.
- Delete the null value in `Description`.
- Change the 'USA' in `Country` column to 'United States'.
- Change the 'UK' and 'U.K.' in `Country` to 'United Kingdom'.
- Delete the negative values in `Quantity`.
- Delete the negative values in `UnitPrice`.

##### To distinguish the dataset after cleaning, we create a new dataframe as 'cleaned_data', making it a copy of 'original_data'. The following cleaning process will be done in 'cleaned_data'.

In [13]:
cleaned_data = original_data.copy()
cleaned_data.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,12/1/2010 8:26,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,12/1/2010 8:26,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,12/1/2010 8:26,3.39,17850.0,United Kingdom


### 5.1 Transfer the data type of `InvoiceDate` to date and time

In [14]:
cleaned_data['InvoiceDate'] = pd.to_datetime(cleaned_data['InvoiceDate'])
cleaned_data['InvoiceDate']

0        2010-12-01 08:26:00
1        2010-12-01 08:26:00
2        2010-12-01 08:26:00
3        2010-12-01 08:26:00
4        2010-12-01 08:26:00
                 ...        
541904   2011-12-09 12:50:00
541905   2011-12-09 12:50:00
541906   2011-12-09 12:50:00
541907   2011-12-09 12:50:00
541908   2011-12-09 12:50:00
Name: InvoiceDate, Length: 541909, dtype: datetime64[ns]

### 5.2 Transfer the data type of `CustomerID` to strings

In [15]:
cleaned_data['CustomerID'] = cleaned_data['CustomerID'].astype(str)
cleaned_data['CustomerID']                                                               

0         17850.0
1         17850.0
2         17850.0
3         17850.0
4         17850.0
           ...   
541904    12680.0
541905    12680.0
541906    12680.0
541907    12680.0
541908    12680.0
Name: CustomerID, Length: 541909, dtype: object

In [16]:
cleaned_data['CustomerID'] = cleaned_data['CustomerID'].str.slice(0, -2)
cleaned_data['CustomerID'] 

0         17850
1         17850
2         17850
3         17850
4         17850
          ...  
541904    12680
541905    12680
541906    12680
541907    12680
541908    12680
Name: CustomerID, Length: 541909, dtype: object

### 5.3 Delete the null value in `Description`

In [17]:
cleaned_data.dropna(subset = ['Description'], inplace = True)

In [19]:
cleaned_data['Description'].isnull().sum()

np.int64(0)

### 5.4 Change the 'USA' in `Country` column to 'United States'

In [21]:
cleaned_data['Country'] = cleaned_data['Country'].replace({'USA': 'United States'})

In [24]:
len(cleaned_data[cleaned_data['Country'] == 'USA'])

0

### 5.5 Change the 'UK' and 'U.K.' in `Country` to 'United Kingdom'

In [22]:
cleaned_data['Country'] = cleaned_data['Country'].replace({'UK' : 'United Kingdom', 'U.K.' : 'United Kingdom'})

In [26]:
print(len(cleaned_data[cleaned_data['Country'] == 'UK']))
print(len(cleaned_data[cleaned_data['Country'] == 'U.K.']))

0
0


### 5.6 Delete the negative values in `Quantity`

In [27]:
cleaned_data = cleaned_data[cleaned_data['Quantity'] >= 0]

In [28]:
len(cleaned_data[cleaned_data['Quantity'] < 0])

0

### 5.7 Delete the negative values in `UnitPrice`

In [30]:
cleaned_data = cleaned_data[cleaned_data['UnitPrice'] >= 0]

In [31]:
len(cleaned_data[cleaned_data['UnitPrice'] < 0])

0

## 6. Save the dataset after cleaning

In [32]:
cleaned_data.to_csv('e_commerce_cleaned.csv', index = False)